### ISSUE TITLE
Clarification for `Accept: */*`

### ISSUE THREAD

**Issue by tobias-tengler** [graphql/graphql-over-http#242]
The spec is pretty clear on what to do if the `Accept` header is not specified or has a value of either `application/graphql-response+json` or `application/json`. It's not so clear to me though, which response `Content-Type` the server should choose in the case of `Accept: */*` (especially during the legacy watershed period). It would be great if the spec could clarify this part.

**Comment by tobias-tengler** [comment]
According to the reference implementation `*/*` is treated as `application/json` for now: https://github.com/graphql/graphql-http/blob/main/src/handler.ts#L689

**Comment by benjie** [comment]
It’s the servers choice; but I think defaulting too application/json prior to watershed and application/graphql-response+json after is reasonable. In fact you could go hybrid and use graphql-response on error even now so you can start using the HTTP codes!

**Comment by Shane32** [comment]
The HTTP protocol has a number of semantics for the `Accept` header, such as q-factor weighting, which are not covered by the GraphQL over HTTP protocol.  A compliant server should fully support the HTTP specification.  I would not think those specifications would need to be reiterated.  It should be rather obvious that `Accept: */*` should act in the same manner as if `Accept` were not supplied, as the GraphQL over HTTP protocol defines the default content type during and after the legacy watershed period, and this matches `Accept: */*`.  Perhaps the wording could be changed to say 'default content type' rather than 'when the Accept header is not specified', just to be more clear.

**Comment by tobias-tengler** [comment]
Yes, that would be good @Shane32.

**Comment by benjie** [comment]
I actually disagree here. A client that sets no accept header expects the default. A client that says `Accept: */*` is explicitly telling the server: use whatever content type is most appropriate, I can handle it.

**Comment by tobias-tengler** [comment]
Yes, that's correct. In reality most web-based GraphQL Clients will actually send `Accept: */*` though, if the Accept header wasn't specified explicitly by the user. See the [fetch spec](https://fetch.spec.whatwg.org/#terminology-headers) (13.)
I don't have numbers to back this up, but I would assume that nearly 100% of GraphQL requests in browsers are done through the `fetch` API and of those at least 90% wouldn't have specified an explicit `Accept` header.

If most legacy web-based GraphQL clients send the `Accept: */*` unknowingly, it would be wrong and breaking IMO to serve them `application/graphql-response+json` prior to watershed. In Hot Chocolate we are handling no Accept header or a value of `*/*` as `application/graphql-response+json` per default. We offer an option to return `appllication/json` for those two cases though, specifically to support these legacy clients up until watershed.

**Comment by Shane32** [comment]
The HTTP RFC 2616 states:

> If no Accept header field is present, then it is assumed that the client accepts all media types.

In other words, a missing `Accept` header is equivalent to `Accept: */*`

**Comment by benjie** [comment]
Though that is indeed what the spec says (which aligns with what I suggest for users who explicitly set `Accept: */*`), I think the majority of GraphQL clients that don’t set Accept and won’t easily be updated to add an Accept header will be one-off shell scripts and other such ad-hoc uses which still expect application/json since that’s the current behaviour of most servers, so maintaining that until watershed makes sense to me.

But either way, it’s the server’s choice, I don’t think we need to dictate this.
